Lib/exception_group.py

```diff
@@ -0,0 +1,215 @@
+
+        excs: sequence of exceptions
+        tb [optional]: the __traceback__ of this exception group.
+        Typically set when this ExceptionGroup is derived from another.
+        """
+        self.excs = excs
+        # self.__traceback__ is updated as usual, but self.__traceback_group__
+        # is set when the exception group is created.
+        # __traceback_group__ and __traceback__ combine to give the full path.
+        self.__traceback__ = tb
+        self.__traceback_group__ = TracebackGroup(self.excs)
+
+    def project(self, condition):
+        """ Split an ExceptionGroup based on an exception predicate
+
+        returns two new ExceptionGroups: match, rest of the exceptions
+        of self for which condition(e) returns True and False, respectively.
+        match and rest have the same nested structure as self, but empty
+        sub-exceptions are not included.
+
+        condition: BaseException --> Boolean
+        """
+        match, rest = [], []
+        for e in self.excs:
+            if isinstance(e, ExceptionGroup): # recurse
+                e_match, e_rest = e.project(condition)
+                if e_match:
+                    match.append(e_match)
+                if e_rest:
+                    rest.append(e_rest)
+            else:
+                if condition(e):
+                    match.append(e)
+                    e_match, e_rest = e, None
+                else:
+                    rest.append(e)
+        return (ExceptionGroup(match, tb=self.__traceback__),
+                ExceptionGroup(rest, tb=self.__traceback__))
+
+    def split(self, type):
+        """ Split an ExceptionGroup to extract exceptions of type E
+
+        type: An exception type
+        """
+        return self.project(lambda e: isinstance(e, type))
+
+    def subgroup(self, keep):
+        """ Split an ExceptionGroup to extract only exceptions in keep
+
+        keep: List[BaseException]
+        """
+        match, _ = self.project(lambda e: e in keep)
+        return match
+
+    def extract_traceback(self, exc):
+        """ returns the traceback of a single exception
+
+        If exc is in this exception group, return its
+        traceback as a list of frames. Otherwise, return None.
+        """
+        if exc not in self:
+            return None
+        result = []
+        e = self.subgroup([exc])
+        while e:
+            tb = e.__traceback__
+            while tb:
+                result.append(tb.tb_frame)
+                tb = tb.tb_next
+            if isinstance(e, ExceptionGroup):
+                assert len(e.excs) == 1 and exc in e
+                e = e.excs[0]
+            else:
+                assert e is exc
+                e = None
+        return result
+
+    def push_frame(self, frame):
+        import types
+        self.__traceback__ = types.TracebackType(
+            self.__traceback__, frame, 0, 0)
+
+    @staticmethod
+    def render(exc, tb=None, indent=0):
```

Comment:
Eventually this needs to be integrated into traceback.py, or at least fit into the same framework.

And I guess we’ll need to teach the C code for printing tracebacks about EGs.